### PR TITLE
Bugfix FXIOS-9528 [TabManagerTests] Fix selectTab causing memory leak errors and remove test sleeps

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9528)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21088)

## :bulb: Description
- Fix memory leak test failures related to selecting tabs
- Added a few new tests
- Small code refactors allowing us to remove sleeps from preexisting TabManagerTests

**Note:** 
To test really thoroughly, you sometimes need to use xcode right click > run repeatedly option to run individual test and/or TabManagerTests suite upwards of 500 times. There is a small (about 1 in 5,000 iterations) chance of `test_selectTab_doesNotRetain` retaining the TabManager. I'm not sure it's worth the extra effort of trying to track this down, as I would guess this is more a test issue rather than an app issue.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift [hunk 1]
```diff
@@ -294,8 +299,8 @@ class TabManagerImplementation: LegacyTabManager, Notifiable, WindowSimpleTabsPr
               let tabID = UUID(uuidString: tab.tabUUID)
         else { return }
 
-        Task {
-            await self.tabSessionStore.saveTabSession(tabID: tabID, sessionData: tabSession)
+        Task { [weak self] in
+            await self?.tabSessionStore.saveTabSession(tabID: tabID, sessionData: tabSession)
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:303]: Wasn't sure... does anyone foresee a case where we WOULD want to retain self here to ensure saving DOES complete? 🤔

Comment (OrlaM) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:317]: This should be fine. A save is triggered so often that even if it does deallocate before it finishes it's likely the previous most recent save is close enough the user won't notice.

**Review by cyndichin — COMMENTED** [review]
@ih-codes thank you for working on these! I just ran it 1000 times and all three tests using `selectTab` fail for me at first run. Given that they aren't passing 💯 , hesistant to merge these in.

image

Also, left a couple of comments that I would like to get your thoughts on. thanks!

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 2]
```diff
@@ -272,7 +272,7 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
 
     // MARK: - Clear and store
     // TODO: FXIOS-7596 Remove when moving the TabManager protocol to TabManagerImplementation
-    func preserveTabs() { fatalError("should never be called") }
+    func preserveTabs(completion: (() -> Void)? = nil) { fatalError("should never be called") }
```
[the remark this answers is not in the document]
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:275]: I've been wondering about these...

[the remark this answers is not in the document]
Comment (OrlaM) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:275]: Feel free to delete any TODO's you find the code that have tickets marked as completed. Unless you can clearly see work is still needed.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 3]
```diff
@@ -469,7 +469,7 @@ class BrowserViewController: UIViewController,
         }
         // Formerly these calls were run during AppDelegate.didEnterBackground(), but we have
         // individual TabManager instances for each BVC, so we perform these here instead.
-        tabManager.preserveTabs()
+        tabManager.preserveTabs(completion: nil)
```
[the remark this answers is not in the document]
Comment (OrlaM) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:472]: This is something with async await that I have not been able to find a satisfying solution to.

**Comment by cyndichin** [comment]
> > @ih-codes thank you for working on these! I just ran it 1000 times and all three tests using `selectTab` fail for me at first run. Given that they aren't passing 💯 , hesistant to merge these in.
> > image
> > Also, left a couple of comments that I would like to get your thoughts on. thanks!
> 
> Huh, they all passed for me with occasionally 1 out of 5000 iterations failing. 🤔 Bitrise also has been happy the 2 times it ran for this PR...
> 
> In your screenshot it looks like you failed on iteration 524, 185, and 987 (not first run)?
> 
> @cyndichin Were you running the whole suite or a specific test, for 1,000 iterations? Can't easily tell that these failures were from a specific test from the screenshot. Anyway, it's not as rare as when I run, but still occurring in less than 0.3% of the time (if I'm reading this right and it's a single test failing 3 out of 1000 iterations).
> 
> Did you try uninstalling the app once before running the tests too? While I was working on this I swear sometimes things behaved differently after a clean install.

Oops I meant first run of 1000 🤣 , should have been more clearer. It was my first attempt in running the tests repeatedly with the value of 1000 times. I was running the `TabManagerTests` specifically and not the whole Unit Tests suite. 

I did not, let me try on a clean build...
....

I ran the tests repeatedly 500 times and I only tried this once to get the following. Here is a better screenshot. It seems the test that failed on 987 run in the first one, probably succeeded but would fail if I ran 1000 😅 .
image

If Bitrise is okay and these tests capture, then I'll lean towards the team to decide if we want to keep this in. We had a discussion about whether proceeding with modern concurrency or not. So we can also hold off until then.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:472]: @ih-codes Just a follow up, discussed this during the team meeting, @OrlaM gave me a better picture regarding these tests. I think next steps for some of the team is to brush up on modern concurrency in which we can have a better discussion if we want to move forward with it as well as Orla investigating.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:275]: @OrlaM There are a number of `FXIOS-7596` TODOs (here's the related QA Verified [JIRA TASK](https://mozilla-hub.atlassian.net/browse/FXIOS-7596), Nov 2023) in this file which seem to suggest all these function definitions should be removed from `LegacyTabManager.swift`. Is that the case or are the TODOs no longer relevant with regards to the tab refactor...?

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ih-codes) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:472]: I agree @cyndichin with what you said above. And like @OrlaM I don't have a great solution off the top of my head.  I wish I didn't have to leave for my MLP training and miss that discussion! It sounds like we should hold off on this PR until we agree on the best path forward.

In a perfect world, `preserveTabs` would be marked `async` and would await all asynchronous tasks before returning. Then we would not need the completion. Any functions calling `perserveTabs` would either `await` `preserveTabs` in turn, or dispatch the work onto another thread (e.g. `Task {... }`). (However, in the latter case, _those_ functions now become the ones hard to test due to what is essentially an asynchronous side effect 😅 ).

Alas this isn't a perfect world... e.g. If you dig into the `perserveTabs` function more, at some point asynchronous work is dispatched (e.g. ` windowManager.performMultiWindowAction` has a throttled `saveSimpleTabs` call) that isn't "awaited" / doesn't have a completion handler.

I definitely have to think on this more, but one possible idea of a sweeping generalization: functions doing async work should either be able to be awaited (for the full scope of their async work to complete)... OR provide a callback, if not. (That said, I can imagine already cases where that would be tedious 😔 )

**Pull request closed without merging**